### PR TITLE
exporter: fix regex for rgw sync metrics

### PR DESCRIPTION
**Pull request by avanthakkar**
Before:
Screenshot from 2024-04-03 15-29-19

After: 
Grafana Dashboard for RGW sync overview now shows data
Screenshot from 2024-03-23 03-12-36

Prometheus metrics fixed
Screenshot from 2024-03-23 03-11-32

Fixes: https://tracker.ceph.com/issues/65091

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by nizamial09** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test api

File: src/exporter/DaemonMetricCollector.cc [hunk 1]
```diff
@@ -335,13 +335,13 @@ DaemonMetricCollector::add_fixed_name_metrics(std::string metric_name) {
   labels_t labels;
   new_metric_name = metric_name;
 
-  std::regex re("^data_sync_from_(.*)\\.");
-    std::smatch match;
-    if (std::regex_search(metric_name, match, re) == true) {
-      new_metric_name = std::regex_replace(metric_name, re, "from_([^.]*)', 'from_zone");
+  std::regex re("data_sync_from_([^_]*)");
```
Comment (jmolmo) [inline src/exporter/DaemonMetricCollector.cc:338]: Why now the metric does not start by "data_sync_from"? I am asking because maybe behind this "innocent" change we have other side-effects and implications.
This happens in all Ceph versions?

**Review by jmolmo — CHANGES_REQUESTED** [review]
I miss completely the explanation about the root cause of this change. And i think it is important to evaluate the impact and possible side effects.
The bug you created seems to be intended only to accomplish only bureaucratic purposes. The only info util is the title itself "exporter: fix regex for rgw sync metrics" 
Please, can you explain what was the root cause?

File: src/exporter/DaemonMetricCollector.cc (hunk 1 above)
Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc:338]: This only happend to Ceph version >= reef & I've added a description in the tracker now, PTAL https://tracker.ceph.com/issues/65091?next_issue_id=65080&prev_issue_id=65110#note-1

File: src/exporter/DaemonMetricCollector.cc [hunk 2]
```diff
@@ -335,13 +335,13 @@ DaemonMetricCollector::add_fixed_name_metrics(std::string metric_name) {
...
-      new_metric_name = std::regex_replace(metric_name, re, "from_([^.]*)', 'from_zone");
-      labels["source_zone"] = quote(match.str(1));
-      return {labels, new_metric_name};
-    }
+  std::regex re("data_sync_from_zone[-_a-zA-Z0-9]*");
+  std::smatch match;
+  if (std::regex_search(metric_name, match, re)) {
+    new_metric_name = std::regex_replace(metric_name, re, "data_sync_from_zone");
+    labels["source_zone"] = quote(match.str(0)); // Using the entire matched substring
+    return {labels, new_metric_name};
```
Comment (pereman2) [inline src/exporter/DaemonMetricCollector.cc:343]: can we test cases to the unit tests?

**Review by jmolmo — CHANGES_REQUESTED** [review]

File: src/exporter/DaemonMetricCollector.cc (hunk 2 above)
Comment (pereman2) [reply src/exporter/DaemonMetricCollector.cc:343]: > I'm already adding some of unit cases here https://github.com/ceph/ceph/pull/54768/files#diff-95c252e20febd13930df478e4ef7e9f0d7f7c202eaf4f9a908f38e8fd28e73e0R1422 Will add some more there itself.

Then finish the other one? I cannot trust this pr without tests

**Review by pereman2 — CHANGES_REQUESTED** [review]

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by jmolmo** [comment]
@avanthakkar : 
The code of the function **does not extract properly** the zone. Please include in this pull request the unit test to have completely clear that we are extracting the zone and the name for the metric properly.
I have proposed you a test function](https://github.com/ceph/ceph/pull/54768/files#discussion_r1560795061). Please get as base for your test/or do another completely new, but I will not approve this pull request until i see a unit test function with **ALL** the rgw sync perf counters tested.

**Comment by avanthakkar** [comment]
> @avanthakkar : The code of the function **does not extract properly** the zone. Please include in this pull request the unit test to have completely clear that we are extracting the zone and the name for the metric properly. I have proposed you a test function](https://github.com/ceph/ceph/pull/54768/files#discussion_r1560795061). Please get as base for your test/or do another completely new, but I will not approve this pull request until i see a unit test function with **ALL** the rgw sync perf counters tested.

Well I've already covered some test cases in unit test PR https://github.com/ceph/ceph/pull/54768/files#diff-95c252e20febd13930df478e4ef7e9f0d7f7c202eaf4f9a908f38e8fd28e73e0R1422 & I just added 1 more unit case in recent commit.  Let me know what more test cases or combination of metric names you'd like me to test.

**Comment by avanthakkar** [comment]
> * **MOVE the UNIT TEST** to this pull request. I cannot see any reason for not doing that.
> 
>     * Test **ALL RGW sync perf counters**

Since there's already a PR in progress for setting up unit tests, I'd rather stick with that one. It's more convenient and since there are already changes there, it makes sense to move forward with it. As for your mention of **ALL**, I'm not quite sure what you mean. Could you please review the tests I've included and suggest any additional test cases? Thanks in advance.

**Comment by jmolmo** [comment]
> I'd rather stick with that one.

@avanthakkar , by the results you obtained until now, it seems your ideas does not seem to work. 
Let me point you some trick, I have used it during the last 30 years without problems. (probably because is the right way to do the things, i didn't invent nothing, just followed the usual, normal, accepted, well tested, and used during years old developing patterns).
"Is a good practice to include in any modification the unit test related to demonstrate that the new function or the modification works as expected. "
In this case, If you do that,  you will verify by yourself that the current implementation does not work.(i am pointing you this small detail during weeks). This is "Why" i am saying that you need to check ALL rgw sync metrics, because as @pereman2 has demonstrated [to you running the test implemented by you and never run by yourself](https://github.com/ceph/ceph/pull/54768#pullrequestreview-2000103007), the extraction of the name of the metric is going to fail with the current implementation. You will need to test all the metrics, because the regex will need to know what are the suffix in each metric.

I will not review more code around this issues until i see the code, the test, and the evidence of the test execution  in this pull request. And for the moment you have my complete opposition to merge this code.

**Comment by avanthakkar** [comment]
@jmolmo @pereman2 I've updated the PR with unit tests for RGW metrics. The regex used mirrors what the Prometheus module previously employed for RGW sync counters, so there weren't many alterations. Regarding the zone name patterns, I've reached out to the RGW team for clarification. However, this shouldn't hinder the PR for now, as we've been using the same regex thus far without any reported issues from Grafana dashboards consuming these metrics.

**Comment by jmolmo** [comment]
@avanthakkar . In first place congratulations!!! Now we have the unit test in the right place.!!! Really very happy with that!!

In second place... unfortunately you have returned to use the regex I pointed you a couple of weeks ago that  is not correct..
_If in the zone name is possible to have a "underscore" character your regex is not going to extract the name of the zone properly_

Please add in the unit test the following metric: ( and make the code work... :-)
**ceph_data_sync_from_A_ZONE_WITHUNDERSCORE_IN_THE_NAME_fetch_bytes**

Probably when you will change the regex to allow "_" in the zone name, you also will understand that in order to produce properly the metric name, is completely needed to know what are ALL the possible suffixes for these kind of metrics.(this is why I asked you to do the test for ALL the real RGW sync perf counters)

Lets go.. we are in right path!!!

**Comment by avanthakkar** [comment]
> @avanthakkar . In first place congratulations!!! Now we have the unit test in the right place.!!! Really very happy with that!!
> 
> In second place... unfortunately you have returned to use the regex I pointed you a couple of weeks ago that is not correct.. _If in the zone name is possible to have a "underscore" character your regex is not going to extract the name of the zone properly_
> 
> Please add in the unit test the following metric: ( and make the code work... :-) **ceph_data_sync_from_A_ZONE_WITHUNDERSCORE_IN_THE_NAME_fetch_bytes**
> 
> Probably when you will change the regex to allow "_" in the zone name, you also will understand that in order to produce properly the metric name, is completely needed to know what are ALL the possible suffixes for these kind of metrics.(this is why I asked you to do the test for ALL the real RGW sync perf counters)
> 
> Lets go.. we are in right path!!!

Yes @jmolmo I understand your perspective. As mentioned in previous comments, there's uncertainty regarding the zone name pattern. Upon reviewing the Prometheus module's implementation, it appears that zone names with underscores aren't permitted. You can find the reference here: [link to PR](https://github.com/ceph/ceph/pull/54768/files#diff-95c252e20febd13930df478e4ef7e9f0d7f7c202eaf4f9a908f38e8fd28e73e0R1422).

Once we receive confirmation from the RGW team regarding any specific zone name patterns, I'll proceed with the necessary modifications. However, until then, I believe it's prudent to rely on the implementation from the Prometheus module.

**Comment by avanthakkar** [comment]
> > @avanthakkar . In first place congratulations!!! Now we have the unit test in the right place.!!! Really very happy with that!!
> > In second place... unfortunately you have returned to use the regex I pointed you a couple of weeks ago that is not correct.. _If in the zone name is possible to have a "underscore" character your regex is not going to extract the name of the zone properly_
> > Please add in the unit test the following metric: ( and make the code work... :-) **ceph_data_sync_from_A_ZONE_WITHUNDERSCORE_IN_THE_NAME_fetch_bytes**
> > Probably when you will change the regex to allow "_" in the zone name, you also will understand that in order to produce properly the metric name, is completely needed to know what are ALL the possible suffixes for these kind of metrics.(this is why I asked you to do the test for ALL the real RGW sync perf counters)
> > Lets go.. we are in right path!!!
> 
> Yes @jmolmo I understand your perspective. As mentioned in previous comments, there's uncertainty regarding the zone name pattern. Upon reviewing the Prometheus module's implementation, it appears that zone names with underscores aren't permitted. You can find the reference here: [link to PR](https://github.com/ceph/ceph/pull/54768/files#diff-95c252e20febd13930df478e4ef7e9f0d7f7c202eaf4f9a908f38e8fd28e73e0R1422).
> 
> Once we receive confirmation from the RGW team regarding any specific zone name patterns, I'll proceed with the necessary modifications. However, until then, I believe it's prudent to rely on the implementation from the Prometheus module.

@jmolmo I've created tracker to include further improvements for zone name (based on info we get from RGW team)
https://tracker.ceph.com/issues/65569

**Comment by jmolmo** [comment]
@avanthakkar , please recover the "promethize" function tests!!..( probably deleted without knowing)

**Review by jmolmo — APPROVED** [review]

File: src/exporter/DaemonMetricCollector.cc [hunk 3]
```diff
@@ -82,35 +84,41 @@ std::string boost_string_to_std(boost::json::string js) {
 
 std::string quote(std::string value) { return "\"" + value + "\""; }
 
-void DaemonMetricCollector::dump_asok_metrics() {
+void DaemonMetricCollector::dump_asok_metrics(bool sort_metrics, int64_t counter_prio,
+                                              bool sockClientsPing, std::string &dump_response,
+                                              std::string &schema_response,
+                                              bool config_show_response) {
```
Comment (pereman2) [inline src/exporter/DaemonMetricCollector.cc:90]: This change is unrelated to this pr...

**Review by pereman2 — CHANGES_REQUESTED** [review]
There are changes from the other pr still here.

**Comment by avanthakkar** [comment]
> There are changes from the other pr still here.

Well as @jmolmo requested to have unit tests for the regex fix introduce in this PR, I had to bring the change from unit test PR here atleast setup else the unit tests won't work. I don't see any other way to introduce the uni-tests in this PR itself

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check
